library/src/main/java/com/novoda/downloadmanager/Wait.java

```diff
@@ -0,0 +1,52 @@
+package com.novoda.downloadmanager;
+
+import android.support.annotation.Nullable;
+
+import com.novoda.notils.logger.simple.Log;
+
+final class Wait {
+
+    private Wait() {
+        // Uses static factory method.
+    }
+
+    static <T> ThenPerform<T> waitFor(@Nullable Object instanceToWaitFor, Object lock) {
+        return new ThenPerform<>(instanceToWaitFor, lock);
+    }
+
+    static class ThenPerform<T> {
+
+        interface Action<T> {
+            T performAction();
+        }
+
+        private final Object instanceToWaitFor;
+        private final Object lock;
+
+        ThenPerform(Object instanceToWaitFor, Object lock) {
+            this.instanceToWaitFor = instanceToWaitFor;
+            this.lock = lock;
+        }
+
+        T thenPerform(Action<T> action) {
```

Comment:
I saw `return null` above because of this method which was used in a case where we don't care about the return type.

Our `Action` here is more like `Callable` from java. And we may make use of `Runnable` in case where we don't care about the output. This means we would have these 2 methods. And we don't have to `return null`  anymore.

```
T thenPerform(Callable<T> callable)

void thenPerform(Runnable runnable)
``

Reply:
Happy to do this in a follow-up PR I am having issues with lint when doing this and I'm not sure why atm. I think it's failing for some reason with the lambdas

Reply:
Sounds good 👍 It is a little improvement imho. It's also fine without it.

Reply:
Ah good point! Forgot they added that incubating feature. d352695